### PR TITLE
rewrite linting

### PR DESCRIPTION
**Pull request by imjasonh**
Instead of `--fail-on-lint-warning` this adds flags for `--lint-require` and `--lint-warn`, which configures lint checks to warn and fail on (currently the default is only to require the new `infodir` linter)

Over time we can add more required checks, starting with them as warnings while we see how noisy that is.

Build configs can `disable` checks (they can no longer `enable` them), in which case a required check becomes a warning instead.

Linting already-built packages has no way to disable a check, since all we have is the package contents, and not the build context.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -15,19 +15,13 @@
...
 
 	"chainguard.dev/melange/pkg/cli"
 )
 
 func main() {
-	ctx, done := signal.NotifyContext(context.Background(), os.Interrupt)
-	defer done()
-
-	if err := cli.New().ExecuteContext(ctx); err != nil {
+	if err := cli.New().Execute(); err != nil {
```
Comment (jonjohnsonjr) [inline main.go:24]: Why drop this?

File: pkg/cli/build.go [hunk 2]
```diff
@@ -211,11 +213,12 @@ func Build() *cobra.Command {
 	cmd.Flags().BoolVar(&debugRunner, "debug-runner", false, "when enabled, the builder pod will persist after the build succeeds or fails")
 	cmd.Flags().BoolVarP(&interactive, "interactive", "i", false, "when enabled, attaches stdin with a tty to the pod on failure")
 	cmd.Flags().BoolVar(&remove, "rm", false, "clean up intermediate artifacts (e.g. container images)")
-	cmd.Flags().BoolVar(&failOnLintWarning, "fail-on-lint-warning", false, "turns linter warnings into failures")
```
Comment (jonjohnsonjr) [inline pkg/cli/build.go:214]: Let's hide this flag with at TODO to drop it (and maybe log a warning if it's true) so we can update `melange` without having to fix the flags up.

File: pkg/cli/lint.go [hunk 3]
```diff
@@ -47,69 +41,32 @@ func Lint() *cobra.Command {
...
-			// Enable non-default lints
-			for _, e := range enabled {
-				if !slices.Contains(disabled, e) {
-					linterSet[e] = struct{}{}
-				}
+			g, ctx := errgroup.WithContext(ctx)
+			g.SetLimit(runtime.GOMAXPROCS(0))
+
+			errs := []error{}
+			var mu sync.Mutex
```
Comment (jonjohnsonjr) [inline pkg/cli/lint.go:48]: I have a slight preference for avoiding the mutex by doing `make([]error, len(args))` and having each goroutine set `errs[i] = err` then letting `errors.Join` filter out the nils.

File: pkg/cli/lint.go [hunk 4]
```diff
@@ -47,69 +41,32 @@ func Lint() *cobra.Command {
...
-
-			o.linters = linters
-
-			return o.RunAllE(ctx, args...)
+			return errors.Join(errs...)
 		},
 	}
 
-	cmd.Flags().StringSliceVar(&enabled, "enabled", []string{}, "enable additional, non-default lints, `--disabled` overrides this")
-	cmd.Flags().StringSliceVar(&disabled, "disabled", []string{}, "disable linters enabled by default or passed in `--enabled`")
```
Comment (jonjohnsonjr) [inline pkg/cli/lint.go:86]: Same comment re: hiding these and logging a warning.

File: pkg/linter/linter.go [hunk 5]
```diff
@@ -512,105 +492,39 @@ func pythonTestPostLinter(_ LinterContext, fsys fs.FS) error {
...
-		if !present {
-			linters = append(linters, l)
-		}
-	}
-
-	return linters
-}
-
-func (lctx LinterContext) lintPackageFs(warn func(error), linters []string, linterClass linter_defaults.LinterClass) error {
+func lintPackageFs(pkgname string, fsys fs.FS, linters []string) error {
```
Comment (jonjohnsonjr) [inline pkg/linter/linter.go:495]: `lintPackageFS` I think is more idiomatic?

File: pkg/linter/linter.go [hunk 6]
```diff
@@ -248,126 +230,126 @@ func srvLinter(_ LinterContext, path string, _ fs.DirEntry) error {
...
+		}
 
-	// No debug sections allowed
-	if file.Section(".debug") != nil || file.Section(".zdebug") != nil {
-		return fmt.Errorf("ELF file is not stripped")
-	}
+		file, err := elf.NewFile(readerAt)
+		if err != nil {
+			// We don't particularly care if this fails otherwise.
+			fmt.Printf("WARNING: Could not open file %q as executable: %v\n", path, err)
```
Comment (jonjohnsonjr) [inline pkg/linter/linter.go:338]: If this is easy enough to disable, I'd rather return an error here than pollute stdout. I think this is a holdover from when we didn't have the check right above that skips non-elf-magic.

File: pkg/linter/linter.go [hunk 7]
```diff
@@ -512,105 +492,39 @@ func pythonTestPostLinter(_ LinterContext, fsys fs.FS) error {
...
-	return lctx.lintPackageFs(warn, linters, linter_defaults.LinterClassBuild)
+	if err := lintPackageFs(packageName, fsys, warn); err != nil {
+		log.Warnf("package linter warning in %s: %v", packageName, err)
+	}
+	return lintPackageFs(packageName, fsys, require)
 }
 
 // Lint the given APK at the given path
-func LintApk(ctx context.Context, path string, warn func(error), linters []string) error {
+func LintApk(ctx context.Context, path string, require, warn []string) error {
```
Comment (jonjohnsonjr) [inline pkg/linter/linter.go:527]: `LintAPK` is idiomatic.

File: pkg/linter/linter.go [hunk 8]
```diff
@@ -512,105 +492,39 @@ func pythonTestPostLinter(_ LinterContext, fsys fs.FS) error {
...
-		}
-	}
-
-	return nil
+	return errors.Join(errs...)
 }
 
 // Lint the given build directory at the given path
-func LintBuild(packageName string, path string, warn func(error), linters []string) error {
+func LintBuild(ctx context.Context, packageName string, path string, require, warn []string) error {
```
Comment (jonjohnsonjr) [inline pkg/linter/linter.go:516]: WDYT about moving the warn vs require distinction outside of this function?

Then we'd just call it twice like:

```go
if err := linter.LintBuild(ctx, pkg, path, warns); err != nil {
    log.Warnf("package linter warning in %s: %v", packageName, err)
}
if err := linter.LintBuild(ctx, pkg, path, requires); err != nil {
  return err
}
```

Or is it better for the require vs warn handling to be the responsibility of this function/package?

Genuinely asking, I haven't thought too deeply about this.

File: pkg/linter/linter.go [hunk 9]
```diff
@@ -18,153 +18,136 @@ import (
...
 	"net/http"
 	"os"
 	"path/filepath"
 	"regexp"
+	"slices"
 	"strings"
 
 	"chainguard.dev/apko/pkg/apk/expandapk"
-	linter_defaults "chainguard.dev/melange/pkg/linter/defaults"
+	"github.com/charmbracelet/log"
```
Comment (jonjohnsonjr) [inline pkg/linter/linter.go:33]: I think we're accidentally pulling this in, see my suggestion below about moving the warn vs require distinction outside of this file.

File: main.go (hunk 1 above)
Comment (imjasonh) [reply main.go:24]: Doing notify handling ourselves broke ctrl+C, just relying on cobra's makes it work again.

File: pkg/cli/build.go (hunk 2 above)
Comment (imjasonh) [reply pkg/cli/build.go:214]: Smart smart.

File: pkg/linter/linter.go (hunk 8 above)
Comment (imjasonh) [reply pkg/linter/linter.go:516]: I don't really love this either. I don't feel strongly, it's only done twice.

File: pkg/cli/lint.go (hunk 3 above)
Comment (imjasonh) [reply pkg/cli/lint.go:48]: Ehhh, that feels clever for a very very small performance gain. If you find out it's slowing us down we can do your thing.

**Review by jonjohnsonjr — APPROVED** [review]

**Pull request merged**